reasoner/planner/ReasonerPlanner.java

```diff
@@ -51,37 +54,71 @@ public ReasonerPlanner(TraversalEngine traversalEng, ConceptManager conceptMgr,
     }
 
     public static ReasonerPlanner create(TraversalEngine traversalEng, ConceptManager conceptMgr, LogicManager logicMgr) {
-        return new GreedyCostSearch.OldPlannerEmulator(traversalEng, conceptMgr, logicMgr);
+        return new RecursivePlanner(traversalEng, conceptMgr, logicMgr);
     }
 
     static Set<Variable> estimateableVariables(Set<Variable> variables) {
         return iterate(variables).filter(Variable::isThing).toSet();
     }
-    public Plan plan(ResolvableConjunction conjunction, Set<Identifier.Variable.Retrievable> bounds) {
-        return planCache.get(new Pair<>(conjunction, bounds), cbPair -> computeResolvableOrdering(logicMgr.compile(cbPair.first()), cbPair.second()));
-    }
 
-    abstract Plan computeResolvableOrdering(Set<Resolvable<?>> resolvables, Set<Identifier.Variable.Retrievable> bounds);
+    static Set<Variable> retrievedVariables(Resolvable<?> resolvable) {
+        return iterate(resolvable.variables()).filter(v -> v.id().isRetrievable()).toSet();
+    }
 
-    boolean dependenciesSatisfied(Resolvable<?> resolvable, Set<Identifier.Variable.Retrievable> bounds, Map<Resolvable<?>, Set<Identifier.Variable.Retrievable>> dependencies) {
+    static boolean dependenciesSatisfied(Resolvable<?> resolvable, Set<Variable> bounds, Map<Resolvable<?>, Set<Variable>> dependencies) {
         return bounds.containsAll(dependencies.get(resolvable));
     }
 
+    public void plan(ResolvableConjunction conjunction, Set<Variable> boundVariables) {
+        CallMode callMode = new CallMode(conjunction, estimateableVariables(boundVariables));
+        // We can't use `planCache.get(Key, Func)` below because of the recursive-update error
+        if (planCache.getIfPresent(callMode) == null) {
+            synchronized (this) {
+                if (planCache.getIfPresent(callMode) == null) {
+                    plan(callMode);
+                }
+            }
+        }
+    }
+
+    public void planAllDependencies(Concludable concludable, Set<Variable> boundVariables) {
+        triggeredCalls(concludable, estimateableVariables(boundVariables), Optional.empty()).forEach(callMode -> plan(callMode.conjunction, callMode.bounds));
```

Comment:
in this case, you do want a line-break before the `.forEach` so it doesn't overrun the line limits